File: Week 2/id_520/94.二叉树的中序遍历.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Stack;

import javax.swing.tree.TreeNode;

/*
 * @lc app=leetcode.cn id=94 lang=java
 *
 * [94] 二叉树的中序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
    //     //方法一，递归
    //     List<Integer> result = new ArrayList<>();
    //     inHelper(root, result);
    //     return result;
    // }
    // public void inHelper(TreeNode root, List<Integer> result) {
    //     if (root == null) return;
    //     inHelper(root.left, result);
    //     result.add(root.val);
    //     inHelper(root.right, result);
    // }
        
        //方法二，非递归的LeetCode最高票代码，前中后序代码相对来说最好记，因为有规律
            LinkedList<Integer> result = new LinkedList<>();
            Stack<TreeNode> stack = new Stack<>();
            while (root != null || !stack.empty()) {
                if (root != null) {
                    stack.push(root);
                    root = root.left;
                } else {
                    TreeNode node = stack.pop();
                    result.add(node.val); //Add after all left children
                    root = node.right;
                }
        }
        return result;
    }
}
// @lc code=end

File: Week 2/id_160/94.二叉树的中序遍历.cpp
/*
 * @lc app=leetcode.cn id=94 lang=cpp
 *
 * [94] 二叉树的中序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> nodes;
        std::stack<TreeNode*> s;
        while(1) {
            while(root) { s.push(root); root=root->left; }
            if(s.empty()) break;
            root=s.top(); s.pop();
            nodes.push_back(root->val);
            root=root->right;
        }
        return nodes;
    }
};

// class Solution {
// public:
//     vector<int> inorderTraversal(TreeNode* root) {
//         stack<TreeNode> s;
//         vector<int> nodes;
//         TreeNode* t = root;
//         while(t || !s.empty()){
//             while(t){  //遍历到最左边的叶结点
//                 s.push(*t);
//                 t = t->left;
//             }
//             if(!s.empty()){
//                 nodes.push_back(s.top().val);
//                 t = s.top().right;
//                 s.pop();
//             }
//         }
//         return nodes;
//     }
// };

// @lc code=end

